fontforge/plugin.c

```diff
@@ -0,0 +1,558 @@
+    } else if (strcasecmp(modestr, "on") == 0) {
+        return sm_on;
+    } else {
+        return sm_ask;
+    }
+}
+
+void *getPluginStartupMode() {
+    return copy(pluginStartupModeString(plugin_startup_mode, true));
+}
+
+void setPluginStartupMode(void *modevoid) {
+    plugin_startup_mode = pluginStartupModeFromString((char *)modevoid);
+}
+
+void SavePluginConfig() {
+    GKeyFile *conf = g_key_file_new();
+    for (GList_Glib *i = plugin_data; i != NULL; i = i->next) {
+        PluginEntry *pe = (PluginEntry *) i->data;
+        if (pe->startup_mode == sm_ask) {
+            continue;    // Don't save merely discovered plugin config
+        }
+        g_key_file_set_string(conf, pe->name, "Package name", pe->package_name);
+        g_key_file_set_string(conf, pe->name, "Module name", pe->module_name);
+        g_key_file_set_string(conf, pe->name, "Active", pluginStartupModeString(pe->startup_mode, false));
+        if (pe->package_url != NULL) {
+            g_key_file_set_string(conf, pe->name, "URL", pe->package_url);
+        }
+    }
+
+    char *pdir = getPluginDirName();
+    if (pdir != NULL) {
+        char *fname = smprintf("%s/plugin_config.ini", pdir);
+        GError *gerror = NULL;
+        g_key_file_save_to_file(conf, fname, &gerror);
+        if (gerror != NULL) {
+            LogError(_("Error saving plugin configuration file '%s': %s\n"), fname, gerror->message);
+            g_error_free(gerror);
+        }
+        free(fname);
+        free(pdir);
+    }
+    g_key_file_free(conf);
+}
+
+static void LoadPluginConfig() {
+    GKeyFile *conf = g_key_file_new();
+    gchar **groups;
+    gsize glen;
+
+    char *pdir = getPluginDirName();
+    if (pdir != NULL) {
+        char *fname = smprintf("%s/plugin_config.ini", pdir);
+        GError *gerror = NULL;
+        g_key_file_load_from_file(conf, fname, G_KEY_FILE_NONE, &gerror);
+        if (gerror != NULL) {
+            if (!g_error_matches(gerror, G_FILE_ERROR, G_FILE_ERROR_NOENT)) {
+                LogError(_("Error reading plugin configuration file '%s': %s\n"), fname, gerror->message);
+            }
+            g_error_free(gerror);
+            gerror = NULL;
+        } else {
+            groups = g_key_file_get_groups(conf, &glen);
+            for (int i = 0; i < glen; ++i) {
+                char *modname = g_key_file_get_string(conf, groups[i], "Module name", NULL);
+                if (modname == NULL) {
+                    LogError(_("No module name for '%s' in plugin config -- skipping.\n"), groups[i]);
+                    continue;
+                }
+                char *pkgname = g_key_file_get_string(conf, groups[i], "Package name", NULL);
+                char *sm_string = g_key_file_get_string(conf, groups[i], "Active", NULL);
+                char *url = g_key_file_get_string(conf, groups[i], "URL", NULL);
+                PluginEntry *pe = NewPluginEntry(copy(groups[i]), pkgname, modname, url, pluginStartupModeFromString(sm_string));
```

Comment:
I can look at pairing with `g_free`.

It seems to me that the two existing calls of `NewPluginEntry()` demonstrate how that's not ideal, given that one copies the candidate `modname` and the other doesn't (because the latter was already generated as a copy). I suppose the "always copy" philosophy does make sense for widely-used interfaces but `NewPluginEntry` is `static`. "Normalizing" such a specific function now seems like speculative generality.

Reply:
Looking at the allocation question further I don't understand the issue, or don't think there is an issue.

A `GList` does its own internal allocations in service of maintaining a list of `data` pointers, which are supplied to the list via `g_list_append` or some other method. The only means of freeing the *data* via those interfaces is with `g_list_clear()` which takes a data-freeing-function as an argument. Otherwise it's your job to manipulate lists in a way that doesn't leak pointed-to heap objects. (See, for example, the construction of the new list and destruction of the old at the end of `PyFF_ConfigurePlugins()`

So I think I'm already following "the rules": The only glib-allocated objects are internal to the `GList` and `GKeyFile` data structures. The former are only ever destroyed with `g_list_free()` and the latter with `g_key_file_free()`.

Reply:
My point was wrt. `g_key_file_get_string`. It allocates a copy, which you should free with `g_free`. But `FreePluginEntry` uses `free`.

Reply:
This is fixed.

Reply:
I also normalized `NewPluginEntry` as requested.